include/v4l2_camera/hysteresis_state_machine.hpp

```diff
@@ -0,0 +1,167 @@
+#ifndef HYSTERESIS_STATE_MACHINE_HPP_
+#define HYSTERESIS_STATE_MACHINE_HPP_
+
+#include <tracetools/utils.hpp>
+
+#include <diagnostic_msgs/msg/diagnostic_status.hpp>
+
+#include <string>
+#include <variant>
+
+namespace custom_diagnostic_tasks
+{
+using DiagnosticStatus_t = unsigned char;
+
+// Helper struct to express state machine nodes
+struct StateBase
+{
+  StateBase(const DiagnosticStatus_t lv) : level(lv), num_observations(1) {}
+
+  DiagnosticStatus_t level;
+  size_t num_observations;
+};
+
+struct Stale : public StateBase
+{
+  Stale() : StateBase(diagnostic_msgs::msg::DiagnosticStatus::STALE) {}
+};
+
+struct Ok : public StateBase
+{
+  Ok() : StateBase(diagnostic_msgs::msg::DiagnosticStatus::OK) {}
+};
+
+struct Warn : public StateBase
+{
+  Warn() : StateBase(diagnostic_msgs::msg::DiagnosticStatus::WARN) {}
+};
+
+struct Error : public StateBase
+{
+  Error() : StateBase(diagnostic_msgs::msg::DiagnosticStatus::ERROR) {}
+};
+
+using StateHolder = std::variant<Stale, Ok, Warn, Error>;
+
+StateHolder generate_state(const DiagnosticStatus_t & state)
```

Comment:
`Function 'generate_state' defined in a header file; function definitions in header files can lead to ODR violations (fix available)clang-tidy[misc-definitions-in-headers](https://clang.llvm.org/extra/clang-tidy/checks/misc/definitions-in-headers.html)`
```suggestion
static StateHolder generate_state(const DiagnosticStatus_t & state)
```
Can be `inline` as well.